astropy/io/fits/tests/test_core.py

```diff
@@ -277,10 +279,24 @@ def make_invalid_hdu():
         hdu = make_invalid_hdu()
         hdu.verify("silentfix+ignore")
 
+        if PYTEST_LT_8_0:
+            ctx_ver_err = ctx_card_n = ctx_zero_idx = nullcontext()
+        else:
+            ctx_ver_err = pytest.warns(
+                fits.verify.VerifyWarning, match="Verification reported errors"
+            )
+            ctx_card_n = pytest.warns(fits.verify.VerifyWarning, match=r"Card \d")
+            ctx_zero_idx = pytest.warns(
+                fits.verify.VerifyWarning,
+                match=r"Note: astropy\.io\.fits uses zero-based indexing",
+            )
+
         # silentfix+warn should be quiet about the fixed HDU and only warn
         # about the unfixable one
         hdu = make_invalid_hdu()
-        with pytest.warns(AstropyUserWarning, match="Illegal keyword name") as w:
+        with ctx_ver_err, ctx_card_n, ctx_zero_idx, pytest.warns(
+            AstropyUserWarning, match="Illegal keyword name"
+        ) as w:
             hdu.verify("silentfix+warn")
```

Comment:
Here, there is a mix of `VerifyWarning` and `AstropyUserWarning`, are they interchangeable? 

Also, why provide a test case that emit all these extra warnings if you don't care about them? Do you really not care if suddenly a new warning gets added or an old warning goes away?

Reply:
`VerifyWarning` is a subclass of `AstropyUserWarning`.
When there is a verification warning, the code will always emit 3 more warnings to explain the issue:
```
WARNING: VerifyWarning: Verification reported errors: [astropy.io.fits.verify]
WARNING: VerifyWarning: Card 6: [astropy.io.fits.verify]
WARNING: VerifyWarning:     Unfixable error: Illegal keyword name 'P.I.' [astropy.io.fits.verify]
WARNING: VerifyWarning: Note: astropy.io.fits uses zero-based indexing.
```
So no need to test those (well maybe once, not sure if they are tested somewhere else).

Reply:
But in any case it's simpler to catch all `VerifyWarning` and then check what's appropriate for the test.

Reply:
Should we also perhaps not raise multiple warnings in the verification for a given issue?

Reply:
@saimn and @astrofrog , why don't you two discuss this, and then one of you can push the changes you desire here? I am not familiar with the tests so I am not comfortable with changing the actual behavior because I don't know the original intent. Thanks!

Reply:
@astrofrog - It's because there may be more than one verification warning, and the first and last line are added to explain things:
```
WARNING: VerifyWarning: Verification reported errors: [astropy.io.fits.verify]
WARNING: VerifyWarning: Card 5: [astropy.io.fits.verify]
WARNING: VerifyWarning:     Card keyword 'test' is not upper case. [astropy.io.fits.verify]
WARNING: VerifyWarning: Card 6: [astropy.io.fits.verify]
WARNING: VerifyWarning:     Card keyword 'p.i.' is not upper case. [astropy.io.fits.verify]
WARNING: VerifyWarning:     Illegal keyword name 'P.I.' [astropy.io.fits.verify]
WARNING: VerifyWarning: Note: astropy.io.fits uses zero-based indexing.
```

Reply:
Is there a consensus?

Reply:
@saimn or @astrofrog , how should we proceed here? My goal is to reduce the number of failures soon. Between this and numpy dev, the dev jobs are a mess now. Please advise. Thanks!

Reply:
Okay, @saimn , I think I addressed your concern for the `io.fits` changes. Please re-review. Thanks!

Reply:
Sorry for the lack of response, sounds good, thanks ;)